src/main/java/com/atlassian/db/replica/internal/circuitbreaker/BreakOnNotSupportedOperations.java

```diff
@@ -1,30 +1,42 @@
 package com.atlassian.db.replica.internal.circuitbreaker;
 
-import com.atlassian.db.replica.api.circuitbreaker.BreakerState;
-import com.atlassian.db.replica.spi.circuitbreaker.CircuitBreaker;
+import com.atlassian.db.replica.api.SqlCall;
 import com.atlassian.db.replica.internal.ReadReplicaUnsupportedOperationException;
+import com.atlassian.db.replica.spi.circuitbreaker.CircuitBreaker;
 
+import java.sql.SQLException;
 import java.sql.SQLFeatureNotSupportedException;
 
-import static com.atlassian.db.replica.api.circuitbreaker.BreakerState.CLOSED;
-import static com.atlassian.db.replica.api.circuitbreaker.BreakerState.OPEN;
-
 public class BreakOnNotSupportedOperations implements CircuitBreaker {
-    private static volatile BreakerState state = CLOSED;
+
+    private static volatile boolean ALL_CALLS_SUPPORTED_SO_FAR = true;
 
     @Override
-    public BreakerState getState() {
-        return state;
+    public boolean canCall() {
+        return ALL_CALLS_SUPPORTED_SO_FAR;
+    }
+
+    @Override
+    public <T> T handle(SqlCall<T> call) throws SQLException {
+        try {
+            return call.call();
+        } catch (ReadReplicaUnsupportedOperationException | SQLFeatureNotSupportedException e) {
+            ALL_CALLS_SUPPORTED_SO_FAR = false;
+            throw e;
+        }
     }
 
     @Override
-    public void handle(Throwable throwable) {
-        if (throwable instanceof ReadReplicaUnsupportedOperationException || throwable instanceof SQLFeatureNotSupportedException) {
```

Comment:
But not without a cost. At the and we always care to handle throwable. The current implementation requires to implement two methods and may grow in the future. The new interface is also more error-prone. The wrong implementation can cut all the calls to DB.

+ no `instanceof`
- internal implementation promotes code duplication,
- harder to implement,
- less stable SPI,
- broader SPI,
- more error-prone,

Reply:
In general, there are more benefits than you listed. Most are in the commit message: f07ba86bc8082f7d2c68cf6d465cbad416a8ec2c

> (-) internal implementation promotes code duplication,

That's also visible in `resilience4j` and anything with callbacks. The `Runnable`/`Callable` split is standard in Java.

> (-) harder to implement,

True! Unless you want to add latency-based CB, or other states, then it's easier.

> (-) less stable SPI,
> (-) broader SPI,

Circuit breakers can have a lot of methods. In `resilience4j` it has 40 instance methods and 14 static methods. Here I just chose a subset (3) that is actually needed by `db-replica`. So it is as narrow as possible. Whatever CB feature is needed by Jira, it is possible with this SPI, so there's less reason to change it, so it's more stable. It can be less stable if `db-replica` needs more from the CB, but that remains to be seen.

> (-) more error-prone,

Old impl can also cut all calls to DB, e.g. by returning the wrong state.